### ISSUE TITLE
send websocket-data after leave of the cbf on_ws_message()   (asynchronous)

### ISSUE THREAD

**Issue by Thalak** [h2o/h2o#572]
Hi,

if I send websocket frames similar to your example:

    static void on_ws_message(h2o_websocket_conn_t *conn, const struct wslay_event_on_msg_recv_arg *arg)
    {
        if (arg == NULL) {
            h2o_websocket_close(conn);
            return;
        }

        if (!wslay_is_ctrl_frame(arg->opcode)) {
            struct wslay_event_msg msgarg = {arg->opcode, arg->msg, arg->msg_length};
            wslay_event_queue_msg(conn->ws_ctx, &msgarg);
        }
    }

it's working fine. The webssocket-recv-cbf and the data send are synchronously
because with leave of  on_ws_message() you call finally the methode:

    h2o_websocket_proceed(conn);

where you call inside 

    wslay_event_send(conn->ws_ctx)

to send the queued data before.

But I have to send data after I have left this function and after you have called  h2o_websocket_proceed(conn).

But If I try to call  'h2o_websocket_proceed(conn)' by my own with the same connection 'conn', gotten by the call of  'on_ws_message()', I get a error with a proken pipe in stream.c at 

    n = write(uv__stream_fd(stream), iov[0].iov_base, iov[0].iov_len);

There seems some dependency between h2o, libuv and wslay. 

Do you know how can I send websocket-data - asynchronous to on_ws_message() with the same websocket connection (channel)?

Thanks

**Comment by kazuho** [comment]
> Do you know how can I send websocket-messages asynchronous to on_ws_message() with the same websocket connection (channel)?

You should call `wslay_event_queue_msg` and then call `h2o_websocket_proceed`.

> I get a error with a proken pipe

The most likely cause of receiving broken pipe is because the peer close the socket.  Can you confirm that it is not the case?

**Comment by Thalak** [comment]
Yes, our client has had a to short timeout and therefore we got the broken pipe error. Now it's working fine.

**Comment by prashanthellina** [comment]
@kazuho If I call `h2o_websocket_proceed` after calling `wslay_event_queue_msg`, will it guarantee that all of the queued data is written to the socket? I ask because the queued messages have buffers which are dynamically allocated in my code and those are no longer valid after the end of the callback function.

If the call to `h2o_websocket_proceed` DOES NOT consume all the data in all the queued messages, how do you suggest that the situation be handled?

Thank you.
